core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -430,8 +517,55 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     }
     newIterator.foreach(addIfAbsent)
     oldIterator.foreach(addIfAbsent)
+    mergedApps
+  }
+
+
+  /**
+   * Scan through all known incomplete applications, and check for any that have been updated.
+   *
+   * After the scan, if there were any updated attempts, [[applications]] is updated
+   * with the new values.
+   *
+   * 1. No attempt to replay the application is made; this scan is a low cost operation.
+   * 2. As this overwrites [[applications]] with a new value, it must not run concurrently
+   * with the main scan for new applications. That is: it must be in the [[checkForLogs()]]
+   * operation.
+   */
+  private[history] def scanAndUpdateIncompleteAttemptInfo(): Unit = {
+    val newAttempts: Iterable[FsApplicationAttemptInfo] = applications
+        .filter( e => !e._2.completed)
+        .flatMap { e =>
+          // build list of (false, attempt) or (true, attempt') values
```

Comment:
comment is still outdated, right?  should be something like "build list of incomplete apps that have been updated"?